ResearchKit/Common/ORKTaskViewController.m

```diff
@@ -893,9 +893,18 @@ - (void)showViewController:(ORKStepViewController *)viewController goForward:(BO
         }
     }
     
-    if (step.identifier && ![_managedStepIdentifiers.lastObject isEqualToString:step.identifier]) {
-        [_managedStepIdentifiers addObject:step.identifier];
+    NSString *stepIdentifier = step.identifier;
+    if (stepIdentifier) {
+        NSUInteger indexOfStep = [_managedStepIdentifiers indexOfObject:stepIdentifier];
+        if (indexOfStep == NSNotFound) {
+            [_managedStepIdentifiers addObject:stepIdentifier];
+        } else {
+            // If the step being shown has been visited before, remove all the managed step identifiers visited after it
+            NSUInteger numberManagedOfStepIdentifiers = _managedStepIdentifiers.count;
+            [_managedStepIdentifiers removeObjectsInRange:NSMakeRange(indexOfStep + 1, numberManagedOfStepIdentifiers - (indexOfStep + 1))];
```

Comment:
Can you explain a little bit more about removing the steps in the loop from `_managedStepIdentifiers`?